logstash-core/src/main/java/org/logstash/instrument/metrics/HistoricMetric.java

```diff
@@ -0,0 +1,230 @@
+
+    // Safeguard for reads against a dead writer
+    private final AtomicReference<Instant> lastCaptured = new AtomicReference<>();
+    private static final Duration STALENESS_THRESHOLD = Duration.ofSeconds(10);
+
+    public HistoricMetric(Metric<T> numeratorMetric, Metric<Long> denominatorMetric) {
+        this(Clock.systemUTC(), numeratorMetric, denominatorMetric);
+    }
+
+    public HistoricMetric(Metric<T> metric) {
+        this(Clock.systemUTC(), metric);
+    }
+
+    HistoricMetric(Clock clock, Metric<T> numeratorMetric, Metric<Long> denominatorMetric) {
+        this.clock = Objects.requireNonNull(clock, "clock must not be null");
+        this.numeratorMetric = Objects.requireNonNull(numeratorMetric, "numeratorMetric must not be null");
+        this.denominatorMetric = Objects.requireNonNullElseGet(denominatorMetric, () -> new UptimeMetric(clock, null));
+    }
+
+    HistoricMetric(Clock clock, Metric<T> metric) {
+        this(clock, metric, null);
+    }
+
+    /**
+     * Capture the current values of the linked metrics
+     * TODO: capturing of an HistoricMetric will need to be scheduled on a repeating cadence
+     */
+    public void capture() {
+        capture(clock.instant(), numeratorMetric.getValue(), denominatorMetric.getValue());
+    }
+
+    void capture(final Instant timestamp, final T numerator, final Long denominator) {
+        final Capture capture = new Capture(timestamp, numerator, denominator);
+        final Capture previousTail = this.tail.getAndSet(capture);
+        previousTail.next.set(capture);
+
+        this.lastCaptured.set(timestamp);
+
+        // rotate our quick references in ascending age order,
+        // bailing entirely if we are too young to warrant continuing
+        for (QuickReference quickReference : quickUpdateReferences) {
+            if (quickReference.rotate(timestamp) == false) { return; }
+        }
+
+        // trim out elements that are no longer necessary
+        compact();
+    }
+
+    // We do not need entries between our earliest capture and the one
+    // referenced by our 15-minutes-ago quick reference, so we compact
+    // our linked list to free the nodes in between.
+    private void compact() {
+        final Capture earliest_capture = quickLifetime.capture.get();
+        if (Objects.isNull(earliest_capture)) { return; }
+
+        final Capture fifteen_minutes_ago = quickFifteenMinutes.capture.get();
+        if (Objects.isNull(fifteen_minutes_ago)) { return; }
+
+        earliest_capture.next.set(fifteen_minutes_ago);
+    }
+
+    /**
+     * Get the available calculated rates, possibly performing a capture if
+     * this {@link HistoricMetric} has become stale.
+     *
+     * @return a {@code Map<String, Double>} of rates made available by this {@link HistoricMetric}.
+     *         Rates that cannot be computed due to insufficient data points will not be included.
+     */
+    public Map<String,Double> availableRates() {
```

Comment:
I've a doubt  to expose a `Map<String, Double>` as result. While it's practical and immediate at first glance, on long term it requires the clients of this API to know how it's structured, and essentially they need how to navigate it to retrieve the needed information. Maybe a custom build `Report` class, that exposes a limited set of methods like `getCurrent()`, `getLast5Minutes()` etc could help better the consumer of this API

Reply:
I have a second draft locally that doubles-down on the `Map<String,Double>`, so that the caller doesn't need to know what rates are available. This enables the historic metric to be set up with user-defined rates. While I am borrowing the current/1m/5m/15m/lifetime from linux load averages, I acknowledge that these particular time-frames may not be super useful to every pipeline and _eventually_ want a user to be able to define the granularity of the rates that are exposed/tracked.